src/repr/scalar/datetime.rs

```diff
@@ -1,176 +1,495 @@
+    /// to None; otherwise generate an error to propagate to the user.
+    pub fn write_field_iff_none(
+        &mut self,
+        f: DateTimeField,
+        u: Option<DateTimeFieldValue>,
+    ) -> Result<(), failure::Error> {
+        use DateTimeField::*;
+
+        match f {
+            Year if self.year.is_none() => {
+                self.year = u;
             }
+            Month if self.month.is_none() => {
+                self.month = u;
+            }
+            Day if self.day.is_none() => {
+                self.day = u;
+            }
+            Hour if self.hour.is_none() => {
+                self.hour = u;
+            }
+            Minute if self.minute.is_none() => {
+                self.minute = u;
+            }
+            Second if self.second.is_none() => {
+                self.second = u;
+            }
+            _ => failure::bail!("{} field set twice", f),
         }
+        Ok(())
     }
+    pub fn check_datelike_bounds(&self) -> Result<(), failure::Error> {
+        if let Some(month) = self.month {
+            if month.unit > 12 || month.unit <= 0 {
+                failure::bail!("MONTH must be (1, 12), got {}", month.unit)
+            }
+        }
+        if let Some(day) = self.day {
+            if day.unit > 31 || day.unit <= 0 {
+                failure::bail!("DAY must be (1, 31), got {}", day.unit)
+            }
+        }
+        if let Some(hour) = self.hour {
+            if hour.unit > 23 || hour.unit < 0 {
+                failure::bail!("HOUR must be (0, 23), got {}", hour.unit)
+            }
+        }
+        if let Some(minute) = self.minute {
+            if minute.unit > 59 || minute.unit < 0 {
+                failure::bail!("MINUTE must be (0, 59), got {}", minute.unit)
+            }
+        }
 
-    for ap in annotated_parts {
-        match ap.fmt {
-            TimePartFormat::SQLStandard(f) => {
-                fill_pdt_sql_standard(&ap.tokens, f, value, &mut pdt)?
+        if let Some(second) = self.second {
+            if second.unit > 60 || second.unit < 0 {
+                failure::bail!("SECOND must be (0, 60), got {}", second.unit)
+            }
+            if second.fraction > 1_000_000_000 || second.fraction < 0 {
+                failure::bail!(
+                    "NANOSECOND must be (0, 1_000_000_000), got {}",
+                    second.fraction
+                )
             }
-            TimePartFormat::PostgreSQL(f) => fill_pdt_pg(&ap.tokens, f, value, &mut pdt)?,
         }
+
+        Ok(())
     }
+    pub fn check_interval_bounds(&self, d: DateTimeField) -> Result<(), failure::Error> {
+        use DateTimeField::*;
 
-    Ok(pdt)
+        match d {
+            Year | Month => {
+                if let Some(month) = self.month {
+                    if month.unit > 12 || month.unit < -12 {
+                        failure::bail!("MONTH must be (-12, 12), got {}", month.unit)
```

Comment:
TIOLI: you've already imported `bail!` so you don't need the `failure::`

Also you could `use failure::error` and these could be `ensure!(month.unit < 13 && month.unit > -13, "MONTH must be (-12, 12), got {}", month.unit)`. Not sure that that would actually end up being smaller/more readable with formatting.

Reply:
Nice. Wasn't aware of this.